src/orion/core/utils/__init__.py

```diff
@@ -110,18 +110,16 @@ def __call__(cls, of_type, *args, **kwargs):
 
         :return: The object which was created on the first call.
         """
-        classes = {clz.__name__.lower(): clz for clz in cls.types}
-
-        if not classes.get(of_type.lower()):
+        if of_type.lower() not in cls.typenames:
             cls.types = list(get_all_subclasses(cls.__base__))
             cls.types = [
                 class_ for class_ in cls.types if class_.__name__ != cls.__name__
             ]
-            classes = {clz.__name__.lower(): clz for clz in cls.types}
+            cls.typenames = list(map(lambda x: x.__name__.lower(), cls.types))
```

Comment:
Do you repeat this in __call__ to enable detection of newly created subclasses after creation of the factory? Seems fine, but I would factor out this method to avoid duplication with __init__. Maybe we should also extend cls.types instead of overwriting it.

Reply:
Also we should add unittests to verify this new behavior.

Reply:
yes.  As the factory could be created at the very beginning when importing some modules before other subclasses are actually defined. Let me see how to make the part code be reusable.  As for overwriting or extending, I think it is the same here, because at the moment if some sub-classes defined before is not in cls.typenames after rewriting, means in the memory we can not find those sub-classes any more, and later creation will cause failure.

Reply:
Yeah, if the code to create the list of types is factored out and reused then overwriting should not be an issue since it's doing the same thing.